### DISCUSSION TITLE
Add new Valve integration

### DISCUSSION THREAD

**Discussion by cibernox** [home-assistant/architecture#975] [Entity Models] [answered]
After initial discussion in https://github.com/home-assistant/architecture/discussions/971, this is a proposal to add a new `Valve` integration.

The proposal is implemented by https://github.com/home-assistant/core/pull/102184

## Problem Statement
Homeowners often use valve-based devices to control the flow of water, gas, or other fluids in their homes. Examples of these devices include water shut-off valves for leak detection and prevention, gas valves for controlling heating systems, and irrigation system valves. While Home Assistant has integrations like `switches` and `covers` that can make-do as valves, that seems like a suboptimal solution. 

Some reasons why this is sub-optimal are:
- Switches are too simple for some kind of valves. Some valves support being partially opened, and have transient states like "opening" and "closing" that switches do not support.
- Covers do support being partially opened and transient states, but some of their features don't apply to valves (e.g. `tilting`)
- Wording for voice assistants often feels off. Valves are opened or closed, but switches are turned on of off.

### Implementation of a valve.

There are several layers of incremental capabilities that valves support

#### States/properties
1. All valves have an `open` and `closed` state.
2. Some valves that are aware of the time it takes them to move may optionally have `opening` and `closing` transient status, leading to `open` and `close` respectively.

#### State attributes
1. Valves that support being partially opened will have some sort of `current_position`, like covers do. If a valve is not fully closed it should be considered open.

#### Services
Assuming a new `valve` namespace is created for the services that affect valves, the services this integration should provide are:
1. `valve.open`, `valve.close` and `valve.toggle`. All valves should respond to these services.
3. `valve.set_position`, which takes a number (integer? float?) representing the % of openess of the valve, 0 being fully closed and 100 being fully opened. 
4. `valve.stop`. Some valves that support opening partially may also support stopping at whatever % of openness they are.

#### Device classes
There may be value in adding device classes to make some valves more human friendly, perhaps assigning them by default more precise icons, ect...
There are two possible ways of approaching this, going very general and vague and maximizing choice.

If we decide to go very general, besides the generic valves (the default), we could have `water`, `gas` and `air` valves. 

If we decide to give many choices, valves can really represents many different kinds of devices. Some that come to mind are:
- shut-off water valves
- Shut-off gas valves
- Sprinklers
- Drip irrigation
- Showers
- Faucets / taps.
- Ventilation
- Bypasses
- Pools and jacuzzis
- Chemical dispensers

#### Others
There will be many existing integrations that will have valves represented as switches today and we cannot expect them to migrate overnight. Some may never do.
Because of that I think that allowing switches to be "aliased" as valves (much like today one can represent switches as light, which hides the switch and creates a virtual `light` entity that proxies the hidden switch) would be very very useful.

**Comment by BradleyFord** [comment]
Good write up

In terms of the choices of the valve types, I would suggest we keep it very
simple to either:
- Valve (Covering all types)
- On/Off or Variable Valve
The user may choose to change from the default MDI icon to something that
fits them better

Other than that the mechanics do not change much from a home assistant
perspective... well that is until you get into 3 way valves but that is a
whole new area and I feel too niche for now.

**Reply by emontnemery** [reply]
We can indeed keep it simple and not add any device classes initially, but instead add them one by one as the need arises.

**Reply by emontnemery** [reply]
On second thought, let's start with device classes for water + gas.

**Comment by emontnemery** [comment]
> Because of that I think that allowing switches to be "aliased" as valves (much like today one can represent switches as light, which hides the switch and creates a virtual light entity that proxies the hidden switch) would be very very useful.

This is a good suggestion 👍 
To be specific it means we should teach the `switch_as_x` integration about `valve` (it already knows about cover, fan, light, lock and siren).

**Comment by cibernox** [comment]
Just to be sure before I jump into implementing this feature, do I have the blessing of the core developers on going ahead with the implementation as it was spec'ed in this conversation? I might have some time tomorrow to kickstart this.

**Reply by emontnemery** [reply, marked as answer]
The proposal is good 👍 
We should start with device classes water + gas in an initial implementation, additional, more specific device classes can be added as needed.

**Reply by jbouwh** [reply]
I doubt about the stop service. I seems odd to me as a value can just be `open`, `closed` or somewhere in between.
It is like a lock. It can have a state opening, but we do not stop it. A value may be could be jammed or in a failed state.

**Reply by cibernox** [reply]
If you think a valve can be partially opened it makes sense. Not all valves are binary (most are tho)

**Reply by jbouwh** [reply]
True, you have the set position service for that. Unlike for `cover`, I do not think we should implement a `stop` service.

**Reply by elupus** [reply]
https://as-schneider.blog/2022/08/10/types-of-valves-used-in-piping-and-how-to-classify-them/

Needle valves can be partially open.

**Reply by jbouwh** [reply]
But what does it mean to `stop` them, and does that make sense?

**Comment by parautenbach** [comment]
I've been working on an irrigation replacement, so I'd be very happy to see this. Thanks. 

Since ESPHome refers to HA device classes in the docs, does that mean these will become available in ESPHome too, or would a separate request/discussion need to be logged?

**Comment by BradleyFord** [comment]
I disagree, I think that we definitely should have a stop service.
This is needed for valves that are not binary

**Comment by BradleyFord** [comment]
If the valve is currently traveling this would request that they stop at
the current position, very similar to ordering blinds to stop whilst they
are moving

**Reply by jbouwh** [reply]
For a cover, I would think it makes sense, not fur a valve. You just set it to a position, and that is enough.

**Comment by BradleyFord** [comment]
Sorry I disagree

It is normal practice for all industrial and commercial valves to have this
capability, I have personally used this functionally many times. (Our
parent company is a global automation and valve positioner vendor)

For example pressing the EStop button (Emergency Stop) it will either force
all valves to stop or move to a safe position.

**Reply by jbouwh** [reply]
An emergency stop would make sense, but still I don't feel we need to add this to the model right now if you ask me. Why would you automate on that. I just don't get it. And if it is for safety I would not let HA deside on that.

**Reply by jbouwh** [reply]
For emergency button's I would say we have a button entity.

**Reply by joostlek** [reply]
I think the biggest difference between a button on such system and on HA is that there is/might be a delay. My curtains work on bluetooth and if I stop my curtains, it has to reconnect and tell it to stop, causing it to stop at an unspecified place.

From a user perspective, what kind of valve would you want to stop at an unspecified place? If you have a valve for your garden hose you probably want to use the specified percentage of open, as you know what the result is going to be, for example 25% is low, 50% medium and 100% is hard. I just can't imagine someone filling their swimming pool with a hose and their phone in their hand with the stream slowly increasing to the point where someone likes the amount of flow going through to click on stop (and maybe have a slight delay because external APIs).

Maybe I am not experienced enough to know other use cases, but I am interested in which scenario this would be used.

**Reply by cibernox** [reply]
I can provide a real use case my father in law would have if he had the sightliest interest in smart stuff. 

He has a waterfall in his garden. A big one. He channeled a spring from the mountains through a big ass pipe to a fake waterfall he built with his excavator. Since the spring's flow is not constant and varies with the season and how much rain they had recently, he often adjusts the valve that controls the flow manually, opening or closing the valve slowly until he's satisfied with the amount of water he's getting.

I think it's not a big deal and if the stop functionality didn't exist. He could make it work by trying specific percentages (30, 35, 40, 45...) until he likes the flow, but nevertheless it's an use case I know off where someone opens a valve and stops when the flow is right.

(We're planning on adding a hydroelectric pelton turbine to get power out of that water, because it is such a waste....)

**Reply by farmio** [reply]
I think a common use case of a valve entity could be heating mixing valves (controlled by some integration that provides PI(D) regulators). 
These, as far as I have seen, usually have absolute positioning only (%) (converted by some controller to travel time or 0-10V signal).

**Comment by BradleyFord** [comment]
So essentially it seems you are suggesting that you move away from 100
years of valve design and operations?
My last comment on this, is that we should follow the standards that are
well proven.

**Comment by MartinHjelmare** [comment]
Please don't use email to respond to this discussion. Email doesn't attach the response to a thread in the discussion page. Please respond in the appropriate thread to keep the conversation structure.

**Comment by emontnemery** [comment]
To finalize the discussion about what services and features are needed, let's list brands which can make use of the valve entity component and what services and features they would need?

Some brands are mentioned in https://github.com/home-assistant/architecture/discussions/971, but the services and features needed for each brand are not listed there.

It's OK to list both integrations which are currently in Home Assistant, and brands which could potentially be added to Home Assistant.

**Reply by emontnemery** [reply]
There have not been a lot of replies to this. If there's no specific reason why the `stop` service is not acceptable, I suggest to go forward with the implementation as suggested, including a `stop service`.

**Comment by cibernox** [comment]
I don't own any valve brand that allows opening partially. I do know that tuya has some valves for radiators that open partially and automatically control the flow of hot water to them to keep the temperature stable, but I believe those are typically not exposed directly as valves but as thermostats.

@BradleyFord do you know of products that do?

**Comment by BradleyFord** [comment]
I use a Tuya valve that has both proportional and an On/Off control

![image](https://github.com/home-assistant/architecture/assets/322815/b82b8028-7418-403b-93db-c97f85f55fa4)

**Comment by emontnemery** [comment]
Some googling suggests valves which can't be set to a certain position but also don't instantly open or close are common. To my understanding the valves are controlled by a motor instead of by a solenoid.

Here's an example https://www.amazon.se/-/en/Electric-Powered-Motorized-Two-Way-Engine/dp/B07DJ78VSR?th=1

For such a valve open + close + stop would make a lot of sense.

**Reply by cibernox** [reply]
That is generally true. My valves are mostly manual "quarter-turn" manual valves motorized with a stepper motor, so while I don't care for stopping them mid-way, it would be technically possible.
